internal/pkg/vm/internal/vmcontext/invocation_context.go

```diff
@@ -0,0 +1,308 @@
+package vmcontext
+
+import (
+	"context"
+
+	"github.com/filecoin-project/go-filecoin/internal/pkg/types"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/vm/abi"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/vm/actor"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/vm/actor/builtin/initactor"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/vm/address"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/vm/internal/exitcode"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/vm/internal/gas"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/vm/internal/gascost"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/vm/internal/message"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/vm/internal/runtime"
+	"github.com/ipfs/go-cid"
+)
+
+type invocationContext struct {
+	rt                *VM
+	msg               internalMessage
+	fromActor         *actor.Actor
+	gasTank           *gas.Tracker
+	isCallerValidated bool
+	allowSideEffects  bool
+	toActor           *actor.Actor
+	stateHandle       internalActorStateHandle
+}
+
+type internalActorStateHandle interface {
+	runtime.ActorStateHandle
+	Validate()
+}
+
+func newInvocationContext(rt *VM, msg internalMessage, fromActor *actor.Actor, gasTank *gas.Tracker) invocationContext {
+	// Note: the toActor and stateHandle are loaded during the `invoke()`
+	return invocationContext{
+		rt:                rt,
+		msg:               msg,
+		fromActor:         fromActor,
+		gasTank:           gasTank,
+		isCallerValidated: false,
+		allowSideEffects:  false,
+	}
+}
+
+type stateHandleContext invocationContext
+
+func (ctx *stateHandleContext) AllowSideEffects(allow bool) {
+	ctx.allowSideEffects = allow
+}
+
+func (ctx *stateHandleContext) LegacyStorage() runtime.LegacyStorage {
+	return ctx.rt.LegacyStorage()
+}
+
+func (ctx *invocationContext) invoke() interface{} {
+	// pre-dispatch
+	// 1. charge gas for message invocation
+	// 2. transfer optional funds
+	// 3. short-circuit _Send_ method
+	// 4. load target actor
+	// 5. load target actor code
+	// 6. create target state handle
+
+	// 1. charge gas for msg
+	ctx.gasTank.Charge(gascost.OnMethodInvocation(&ctx.msg))
+
+	// 2. transfer funds carried by the msg
+	ctx.rt.transfer(ctx.msg.from, ctx.msg.to, ctx.msg.value)
```

Comment:
Good catch. Fixed. 

Re: have the to resolved outside. I had it like that originally, but then the "Send" needs it too, so I moved it inside the invoke which is common code to top level messages and sub messages. Since it is now abstracted to a `resolveTarget` method, we could have that call duplicated in both execution paths. 

Re: asserting on address. Sure thing.

Reply:
Ok, assertions are fine.